File: src/PrgData/LogRequestHelper.vb
Imports log4net
Imports System.IO
Imports System.Net.Mail
Imports System.Configuration
Imports PrgData.Common


Public Class LogRequestHelper

	Private Shared _Logger As ILog = LogManager.GetLogger(GetType(LogRequestHelper))

	Public Shared Function NeedLogged() As Boolean
		Return Convert.ToBoolean(ConfigurationManager.AppSettings("MustLogHttpcontext"))
	End Function

	Public Shared Sub ForceMailWithRequest(ByVal MessageText As String, ByVal exception As Exception)
		InternalMailWithRequest(MessageText, exception)
	End Sub

	Public Shared Sub MailWithRequest(ByVal Logger As ILog, ByVal MessageText As String, ByVal exception As Exception)
		'Если логгер не установлен, то будем использовать логгер LogRequestHelper
		Dim l = Logger
		If l Is Nothing Then l = _Logger
		'Если exception не установлен, то будем брать исключение от сервера
		Dim ex = exception
		If ex Is Nothing Then ex = HttpContext.Current.Server.GetLastError()
		If NeedLogged() Then
			InternalMailWithRequest(MessageText, exception)
		Else
			l.Error(MessageText, exception)
		End If
	End Sub

	Private Shared Sub InternalMailWithRequest(ByVal MessageText As String, ByVal exception As Exception)
		Try
			Dim tmpRequestFileName As String = Path.GetTempFileName()
			HttpContext.Current.Request.SaveAs(tmpRequestFileName, True)
			Try
				_Logger.Error(MessageText, exception)

				If exception IsNot Nothing Then
					MessageText = String.Format("{0} " & vbCrLf & "{1}", MessageText, exception)
				End If
				MessageText = String.Format( _
					"Date: {0}" & vbCrLf & _
					"User: {1}" & vbCrLf & _
					"{2}", _
					DateTime.Now, _
					ServiceContext.GetUserName(), _
					MessageText)

				Dim httpRequestContent = File.ReadAllText(tmpRequestFileName)

				MailHelper.Mail(MessageText, Nothing, httpRequestContent, "HTTPReguest.txt")

			Finally
				Try
					File.Delete(tmpRequestFileName)
				Catch ex As Exception
					_Logger.Error("Ошибка при удалении временного файла для хранения HTTP-запроса", ex)
				End Try
			End Try

		Catch err As Exception
			_Logger.Error("Ошибка в MailWithRequest", err)
		End Try
	End Sub

End Class

File: src/PrgData.Common/LogRequestHelper.cs
using System;
using System.Web;
using log4net;
using System.IO;
using System.Net.Mail;
using System.Configuration;
using PrgData.Common;

namespace PrgData
{
	public class LogRequestHelper
	{
		protected static ILog _Logger = LogManager.GetLogger(typeof(LogRequestHelper));

		public static bool NeedLogged()
		{
			return Convert.ToBoolean(ConfigurationManager.AppSettings["MustLogHttpcontext"]);
		}

		public static void MailWithRequest(ILog logger, string messageText, Exception exception)
		{
			//Если логгер не установлен, то будем использовать логгер LogRequestHelper
			var l = logger ?? _Logger;
			//Если exception не установлен, то будем брать исключение от сервера
			var ex = exception ?? HttpContext.Current.Server.GetLastError();
			if (NeedLogged())
				MailWithRequest(messageText, ex);
			else
				l.Error(messageText, ex);
		}

		public static void MailWithRequest(string text, Exception exception = null)
		{
			try {
				var tmpRequestFileName = Path.GetTempFileName();
				HttpContext.Current.Request.SaveAs(tmpRequestFileName, true);

				try {
					_Logger.Error(text, exception);

					if (exception != null)
						text = String.Format("{0} " + Environment.NewLine + "{1}", text, exception);
					text = String.Format(
						"Date: {0}" + Environment.NewLine +
							"User: {1}" + Environment.NewLine +
							"{2}",
						DateTime.Now,
						ServiceContext.GetUserName(),
						text);

					var httpRequestContent = File.ReadAllText(tmpRequestFileName);

					MailHelper.Mail(text, null, httpRequestContent, "HTTPReguest.txt");
				}
				finally {
					try {
						File.Delete(tmpRequestFileName);
					}
					catch (Exception ex) {
						_Logger.Error("Ошибка при удалении временного файла для хранения HTTP-запроса", ex);
					}
				}
			}
			catch (Exception err) {
				_Logger.Error("Ошибка в MailWithRequest", err);
			}
		}
	}
}